### PR TITLE
new endpoint for speakers checkins stats

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3325540&stafilter=NotComplete&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php [hunk 1]
```diff
@@ -435,6 +438,74 @@ public function getAllSummitByIdOrSlugRegistrationStats($id)
...
     }
 
+    /**
+     * @param $id
+     * @return mixed
+     */
+    public function getAttendeesCheckinsOverTimeStats($id)
+    {
+        return $this->processRequest(function () use ($id) {
+            $summit = $this->repository->getById(intval($id));
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:448]: there is an strategy for this

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php [hunk 2]
```diff
@@ -435,6 +438,74 @@ public function getAllSummitByIdOrSlugRegistrationStats($id)
...
+            if (is_null($filter)) $filter = new Filter();
+
+            $filter->validate([
+                'start_date' => 'sometimes|required|date_format:U',
+                'end_date' => 'sometimes|required_with:start_date|date_format:U|after:start_date',
+            ]);
+
+            return $this->ok
+            (
+                SerializerRegistry::getInstance()
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:491]: send paginated response envelope

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php [hunk 3]
```diff
@@ -435,6 +438,74 @@ public function getAllSummitByIdOrSlugRegistrationStats($id)
...
+                return $this->error403(['message' => sprintf("Member %s has not permission for this Summit", $current_member->getId())]);
+
+            $group_by = Request::get('group_by');
+            if(!in_array($group_by, IStatsConstants::AttendeesCheckinsAllowedGroupBy))
+                throw new ValidationException("invalid group by");
+
+            $page = 1;
+            $per_page = PaginationValidationRules::PerPageMin;
+
+            if (Request::has('page')) {
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:465]: lets create a new trait
ParseAndGetPaginationParams
like
ParseAndGetFilter
and lets encapsulate all this logic there
replace here https://github.com/OpenStackweb/summit-api/blob/c52137708ca99a0eb48036869bd75fe9e79f7091/app/Http/Controllers/Apis/Protected/Summit/Traits/ParametrizedGetAll.php#L105 too
method suggestion
static public function getPaginationParams():array{
    // default values
            $page = 1;
            $per_page = is_null($defaultPageSize) ? PaginationValidationRules::PerPageMin : call_user_func($defaultPageSize);

            if (Request::has('page')) {
                $page = intval(Request::get('page'));
            }

            if (Request::has('per_page')) {
                $per_page = intval(Request::get('per_page'));
            }

return [$page, $per_page];
}

and the using the trait u could do

use ParseAndGetPaginationParams;

// somewhere bellow

list($page, $per_page) = self::getPaginationParams();

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments much thanks

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php (hunk 1 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:448]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php [hunk 4]
```diff
@@ -435,6 +438,74 @@ public function getAllSummitByIdOrSlugRegistrationStats($id)
...
+                $page = intval(Request::get('page'));
+            }
+
+            if (Request::has('per_page')) {
+                $per_page = intval(Request::get('per_page'));
+            }
+
+            $filter = null;
+
+            if (Request::has('filter')) {
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:475]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php (hunk 2 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:491]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php (hunk 3 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php]: done

**Comment by romanetar** [comment]
@smarcet  all review comments resolved

File: app/Models/Foundation/Summit/SummitRegistrationStats.php [hunk 5]
```diff
@@ -645,4 +650,77 @@ public function getAttendeesCheckinPerBadgeFeatureType(?DateTime $startDate  = n
...
+    /**
+     * @param string $groupBy
+     * @param int $page
+     * @param int $per_page
+     * @param DateTime|null $startDate
+     * @param DateTime|null $endDate
+     * @return PagingResponse
+     */
+    public function getAttendeesCheckinsGroupedBy(
+        string $groupBy, int $page, int $per_page, ?DateTime $startDate  = null, ?DateTime $endDate = null): PagingResponse
```
Comment (smarcet) [inline app/Models/Foundation/Summit/SummitRegistrationStats.php:662]: please check DoctrineSpeakerRepository::getSpeakerBYSummitAndOnSchedule as use PagingInfo Util Class
https://github.com/OpenStackweb/summit-api/blob/c52137708ca99a0eb48036869bd75fe9e79f7091/app/Repositories/Summit/DoctrineSpeakerRepository.php#L608

File: app/Models/Foundation/Summit/SummitRegistrationStats.php [hunk 6]
```diff
@@ -645,4 +650,77 @@ public function getAttendeesCheckinPerBadgeFeatureType(?DateTime $startDate  = n
...
+            return new PagingResponse
+            (
+                $total,
+                $per_page,
+                $page,
+                intval(ceil($total / $per_page)),
+                $res
+            );
+        } catch (\Exception $ex) {
+
```
Comment (smarcet) [inline app/Models/Foundation/Summit/SummitRegistrationStats.php:715]: please log exception as error

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php [hunk 7]
```diff
@@ -435,6 +441,61 @@ public function getAllSummitByIdOrSlugRegistrationStats($id)
...
+
+            if (is_null($summit)) return $this->error404();
+
+            $current_member = $this->resource_server_context->getCurrentUser();
+            if (!is_null($current_member) && !$current_member->isAdmin() && !$current_member->hasPermissionFor($summit))
+                return $this->error403(['message' => sprintf("Member %s has not permission for this Summit", $current_member->getId())]);
+
+            $group_by = Request::get('group_by');
+            if(!in_array($group_by, IStatsConstants::AttendeesCheckinsAllowedGroupBy))
+                throw new ValidationException("invalid group by");
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:463]: please inform on exception message which are the valid groups by

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Models/Foundation/Summit/SummitRegistrationStats.php (hunk 6 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/SummitRegistrationStats.php:715]: done

File: app/Models/Foundation/Summit/SummitRegistrationStats.php (hunk 5 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/SummitRegistrationStats.php:662]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php [hunk 8]
```diff
@@ -435,6 +441,61 @@ public function getAllSummitByIdOrSlugRegistrationStats($id)
...
+                    return [
+                        'start_date' => 'sometimes|required|date_format:U',
+                        'end_date' => 'sometimes|required_with:start_date|date_format:U|after:start_date',
+                    ];
+                });
+
+            $start_date = null;
+            $end_date = null;
+
+            if($filter->hasFilter('start_date')){
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:484]: These are logic portions from different stats. There isn't dup logic

[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitApiController.php:484]: done

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**